runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/ConfigureTask.java

```diff
@@ -158,8 +172,86 @@ else if ("http".equals(configURL.getProtocol()) || "https".equals(configURL.getP
+
+            configWatcherRef = commonPool().submit(() -> watchConfig(rootNamespace));
+        }
+        return null;
+    }
+
+    public void close()
+    {
+        if (configWatcherRef != null)
+        {
+            configWatcherRef.cancel(true);
+        }
+    }
+
+    private void watchConfig(
+        NamespaceConfig rootNamespace)
+    {
+        try
+        {
+            WatchService watchService;
+            Path configPath = Paths.get(new File(configURL.getPath()).getAbsolutePath());
+
+            watchService = FileSystems.getDefault().newWatchService();
+            configPath.getParent().register(watchService, ENTRY_MODIFY, ENTRY_CREATE, ENTRY_DELETE);
+
+            Path configFileName = configPath.getFileName();
+            while (true)
+            {
+                if (rootNamespace != null)
+                {
+                    try
+                    {
+                        final WatchKey key = watchService.take();
+                        // Sleep is needed to prevent receiving two separate ENTRY_MODIFY events:
+                        // file modified and timestamp updated.
+                        // Instead, receive one ENTRY_MODIFY event with two counts.
+                        Thread.sleep(50);
+
+                        for (WatchEvent<?> event : key.pollEvents())
+                        {
+                            final Path changed = (Path) event.context();
+                            if (changed.equals(configFileName))
+                            {
+                                CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+                                for (DispatchAgent dispatcher : dispatchers)
+                                {
+                                    future = CompletableFuture.allOf(future, dispatcher.detach(rootNamespace));
+                                }
+                                future.join();
+                                String configText;
+                                try (InputStream input = configURL.openConnection().getInputStream())
+                                {
+                                    configText = new String(input.readAllBytes(), UTF_8);
+                                }
+                                catch (IOException ex)
+                                {
+                                    configText = CONFIG_TEXT_DEFAULT;
+                                }
+                                rootNamespace = configure(configText);
+                            }
+                        }
+                        key.reset();
+                    }
+                    catch (InterruptedException ex)
+                    {
+                        watchService.close();
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            rethrowUnchecked(ex);
         }
+    }
```

Comment:
This method is fairly unwieldy and trying to handle a lot of different cases as if it was a linear flow.

I think we might need to restructure the code into 3 tasks.
 - watcher task
 - register task
 - unregister task

The watcher task is responsible for detecting a new configuration, either file based or http based, so initially `FileWatcherTask`, later `HttpWatcherTask`.

When the `FileWatcherTask` detects a file delete or file modify event on the config file it is watching, it submits an `UnregisterTask` to unregister the namespace, detaching any streams for bindings in the namespace.

When the `FileWatcherTask` detects a file create or file modify event on the config file it is watching, it submits a `RegisterTask` to register the new namespace.

I think this design will simplify the approach into more bite-sized chunks of functionality, and also lend it self to being extended for HttpWatcherTask via long-polling, while keeping the register and unregister behaviors unaffected.

Reply:
It made the code much simpler. I was also thinking of doing something similar, but wanted to get your opinion first (https://github.com/aklivity/zilla/pull/141#issuecomment-1375548091).